### PR TITLE
[feat] Add metrics exporter and dashboard for TransferQueue

### PR DESCRIPTION
**Pull request by RobotGF**
Expose Prometheus metrics from the controller and storage units so TransferQueue activity can be monitored end to end. Include a Grafana dashboard and tests so the observability workflow is easier to validate and adopt.

### CODE REVIEW

**Comment by RobotGF** [comment]
截屏2026-04-23 17 29 53
used to add Prometheus and  Grafana for RL training

**Comment by 0oshowero0** [comment]
Thanks for your contribution! You can add my wechat through the community wechat group QR code (posted on README) to further discuss this PR.

File: transfer_queue/storage/simple_backend.py [hunk 1]
```diff
@@ -70,6 +71,32 @@ def __init__(self, storage_size: int):
...
+        try:
+            import torch
+
+            if isinstance(val, torch.Tensor):
+                return val.nelement() * val.element_size()
+        except ImportError:
+            pass
+        if hasattr(val, "nbytes"):
+            return val.nbytes
+        return 0
```
Comment (0oshowero0) [inline transfer_queue/storage/simple_backend.py]: In real-world scenario there will be many non-tenor data. Can we directly fetch the memory usage by the process memory occupancy? It will be more straightforward.

Comment (0oshowero0) [reply transfer_queue/storage/simple_backend.py]: It also helps to reduce the metric collection overhead

File: transfer_queue/storage/simple_backend.py [hunk 2]
```diff
@@ -294,12 +331,17 @@ def _worker_routine(self) -> None:
                     if operation == ZMQRequestType.PUT_DATA:  # type: ignore[arg-type]
                         with perf_monitor.measure(op_type="PUT_DATA"):
                             response_msg = self._handle_put(request_msg)
+                        self._op_counts["PUT_DATA"] += 1
```
Comment (0oshowero0) [inline transfer_queue/storage/simple_backend.py]: Can we extract these statistics leveraging existing `IntervalPerfMonitor`'s capability?

File: transfer_queue/config.yaml [hunk 3]
```diff
@@ -1,6 +1,11 @@
 # This is the default configuration of TransferQueue. Users may modify the default value
 # and use transfer_queue.init(conf) to overwrite the config entries.
 
+# Prometheus metrics exporter.
+# Can also be enabled via the TQ_METRICS_ENABLED=true environment variable.
```
Comment (0oshowero0) [inline transfer_queue/config.yaml]: It's better only use config to control whether metrics is enabled

File: transfer_queue/controller.py [hunk 4]
```diff
@@ -2135,3 +2143,33 @@ def register_sampler(
...
+        to call multiple times -- subsequent calls return the existing endpoint.
+
+        Returns:
+            The metrics endpoint address in ``host:port`` format.
+        """
+        if self._metrics is not None:
+            return self._metrics_endpoint
+        from transfer_queue.metrics import TQMetricsExporter
+
+        self._metrics = TQMetricsExporter(self)
```
Comment (0oshowero0) [inline transfer_queue/controller.py]: We may need to decouple the metric collection from `TransferQueueController`, since it has to serve the whole cluster's request and it's likely to become a single-point bottleneck.

File: transfer_queue/metrics.py [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+        self._storage_unit_infos: dict[str, ZMQServerInfo] = {}
+        self._zmq_ctx: zmq.Context | None = None
+        self._zmq_sockets: dict[str, zmq.Socket] = {}
+        self._known_partition_ids: set[str] = set()
+        self._known_consumption_labels: set[tuple[str, str]] = set()
+        self._metrics_endpoint: str = ""
+        self.registry = CollectorRegistry()
+        self._define_metrics()
+
+    def _define_metrics(self) -> None:
```
Comment (0oshowero0) [inline transfer_queue/metrics.py:94]: We can discuss about these metrics later, and only preserve essential ones for this PR.

File: transfer_queue/config.yaml (hunk 3 above)
Comment (ji-huazhong) [reply transfer_queue/config.yaml]: > It's better only use config to control whether metrics is enabled最好只使用配置来控制是否启用指标。
+1

File: transfer_queue/config.yaml [hunk 6]
```diff
@@ -1,6 +1,10 @@
 # This is the default configuration of TransferQueue. Users may modify the default value
 # and use transfer_queue.init(conf) to overwrite the config entries.
 
+# Prometheus metrics exporter.
```
Comment (0oshowero0) [inline transfer_queue/config.yaml:4]: Please also update the config in ./scripts/performance_test

File: transfer_queue/metrics.py [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+
+from transfer_queue.utils.zmq_utils import (
+    ZMQMessage,
+    ZMQRequestType,
+    ZMQServerInfo,
+    create_zmq_socket,
+    format_zmq_address,
+)
+
+logger = logging.getLogger(__name__)
```
Comment (0oshowero0) [inline transfer_queue/metrics.py]: logger has been simplified by https://github.com/Ascend/TransferQueue/pull/84. Now we can use the following lines instead:

```python3
from transfer_queue.utils.logging_utils import get_logger

logger = get_logger(__name__)
```

File: transfer_queue/metrics.py [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+logger.setLevel(os.getenv("TQ_LOGGING_LEVEL", logging.INFO))
+
+if not logger.hasHandlers():
+    handler = logging.StreamHandler()
+    handler.setFormatter(logging.Formatter("%(asctime)s - %(levelname)s - %(name)s - %(message)s"))
+    logger.addHandler(handler)
+
+TQ_METRICS_PORT = int(os.environ.get("TQ_METRICS_PORT", 0))
```
Comment (0oshowero0) [inline transfer_queue/metrics.py]: I suggest using config to set metric ports, and preserve the other two as env var.

Comment (0oshowero0) [reply transfer_queue/metrics.py]: TQ_METRICS_PORT-> TQ_METRICS_EXPORTER_PORT?

File: transfer_queue/metrics.py [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+        When ``TQ_METRICS_PORT`` is ``0`` (the default), the OS assigns a free
+        port automatically — the actual port is read back from the server socket.
+
+        Args:
+            node_ip: The IP address of the node hosting the Controller Actor.
+
+        Returns:
+            The metrics endpoint address in ``host:port`` format.
+        """
+        httpd, _ = start_http_server(TQ_METRICS_PORT, registry=self.registry)
```
Comment (0oshowero0) [inline transfer_queue/metrics.py]: In newer version of prometheus_client (e.g., 0.17.1), `start_http_server` returns nothing. Please align with new behaviour of this function

File: transfer_queue/storage/simple_storage.py [hunk 10]
```diff
@@ -478,6 +490,31 @@ def _handle_clear(self, data_parts: ZMQMessage) -> ZMQMessage:
...
+            process_rss = psutil.Process().memory_info().rss
+        except Exception:
+            process_rss = 0
+
+        metrics = {
+            "storage_unit_id": self.storage_unit_id,
+            "capacity": self.storage_unit_size,
+            "active_keys": self.storage_data.active_key_count,
+            "process_rss_bytes": process_rss,
+        }
```
Comment (0oshowero0) [inline transfer_queue/storage/simple_storage.py:513]: Can we also add raw ops to the metrics, so we can know the process time for each operation's (PUT_DATA, GET_DATA, CLEAR_DATA) request rate, request latency P50/P99?

File: transfer_queue/storage/simple_storage.py [hunk 11]
```diff
@@ -478,6 +490,83 @@ def _handle_clear(self, data_parts: ZMQMessage) -> ZMQMessage:
...
+        """
+        total = cumulative_counts[-1] if cumulative_counts else 0
+        if total == 0:
+            return 0.0
+        target = q * total
+        prev_bound = 0.0
+        prev_cumulative = 0.0
+        for bound, cum_count in zip(hist._upper_bounds, cumulative_counts):
+            if cum_count >= target:
+                fraction = (target - prev_cumulative) / (cum_count - prev_cumulative) if cum_count > prev_cumulative else 0
```
Comment (0oshowero0) [inline transfer_queue/storage/simple_storage.py]: We can move these util functions in `metrics.py` in the future

**Review by 0oshowero0 — APPROVED** [review]
Great work! Thanks for the contribution~
